### DISCUSSION TITLE
Easier OTP import

### DISCUSSION THREAD

**Discussion by Quintasan** [zeapo/Android-Password-Store#1528] [Ideas]
I have a plenty of OTP keys in andOTP which I would like to import to pass but manually typing in the otpauth:// string on a phone is tedious and error-prone and the Add OTP button only offers scanning a QR code which is physically impossible. I know that andOTP can show you the QR code and you can save it as an image but that doesn't help since you can't use an image.

My two ideas
- allow using an image file
- show a dialog window for putting in just the secret

I'm aware that there can be different algorithms for generating the OTP and can have more digits but as far as I'm aware 60 second expiry, 6 digits and SHA1 is the de-facto standard.

I did dabble in Android development so I could try contributing this feature but I'd like to ask if there's interest in it beforehand.

**Comment by msfjarvis** [comment]
We already do allow importing by providing just the secret, but it's entirely possible it's not part of the public release given how long it's been since we did one. I'd take a PR to import from an image.

**Reply by Quintasan** [reply]
>28 Jul 2021

Oh, I never noticed there was no new release, since everything was working A-okay for me. Actually that I think about it's pretty awesome, so thanks for your hard work @msfjarvis!

A new release would be awesome, but I imagine there are some reasons for this, so I'm not going to ask for one.

>I'd take a PR to import from an image.

It's been a big while since I did any Android development, so I imagine it's going to take me a while to set things up but I'll try putting something together over the weekend.

**Reply by msfjarvis** [reply]
> > 28 Jul 2021
> 
> Oh, I never noticed there was no new release, since everything was working A-okay for me. Actually that I think about it's pretty awesome, so thanks for your hard work @msfjarvis!

Super glad to hear that :-)

> 
> A new release would be awesome, but I imagine there are some reasons for this, so I'm not going to ask for one.

I appreciate your being considerate. The reasons are not all that secret, we have large scale rewrites ongoing to replace OpenKeychain (#1523) and there's another impending one regarding file system access. These need to reach a certain level of done-ness before we are in a state where a release can be considered. 

> 
> > I'd take a PR to import from an image.
> 
> It's been a big while since I did any Android development, so I imagine it's going to take me a while to set things up but I'll try putting something together over the weekend.

We have some things written down that can help you get started with this particular codebase in the [contributing guidelines](https://github.com/android-password-store/Android-Password-Store/blob/develop/CONTRIBUTING.md), and I'll put down some notes to help you get started with this particular issue.


#### Implementation notes

We use a fork of the [zxing-android-embedded](https://github.com/journeyapps/zxing-android-embedded) library for doing QR recognition from the camera, which in turn uses [ZXing](https://github.com/zxing/zxing) to parse barcodes. [This](https://stackoverflow.com/a/29650479) StackOverflow answer details how to decode a QR code from an image using ZXing. To obtain the image, use the [GetContent](https://developer.android.com/reference/androidx/activity/result/contract/ActivityResultContracts.GetContent) contract with a `image/*` MIME and follow the documentation there to obtain the raw data and construct a Bitmap from.
